### PR TITLE
Add description for behavior-keys file to README

### PR DESCRIPTION
**Pull request by TimDiekmann**
## 🌟 What is the purpose of this PR?

Behaviors are accompanied by a .json file that defines the types of the fields used within the behavior. This needs to be described in the README.

## 🔗 Related links

- [Asana task description](https://app.asana.com/0/1199550852792314/1201494000207946/f)

### CODE REVIEW

File: packages/engine/README.md [hunk 1]
```diff
@@ -48,29 +49,28 @@ The [HASH glossary](https://hash.ai/glossary?utm_medium=organic&utm_source=githu
...
+As outlined above, this project is the next-generation of our simulation engine, and differs from the one currently powering [hCore](https://hash.ai/platform/core?utm_medium=organic&utm_source=github_readme_engine) and [hCloud](https://hash.ai/platform/cloud?utm_medium=organic&utm_source=github_readme_engine). It's published here as a pre-release technology preview, and as such the feature-set and codebase should be considered unstable until it's released. That means that there are a number of features you may use on the HASH platform that at present may not be supported by this project, notably:
+* Python runners, and therefore **Python behaviors** are currently **disabled**. This is a high-priority item for us and will be one of the main items of development focused on in the near future. Much of the implementation is finished and in this repository should explore it be of interest (Although as it is _not_ completely finished, expect to find bugs).
```
Comment (Alfred-Mountfield) [inline packages/engine/README.md:55]: Your IDE lied to you I'm afraid. Exploring it is the correct tense here

File: packages/engine/README.md [hunk 2]
```diff
@@ -48,29 +49,28 @@ The [HASH glossary](https://hash.ai/glossary?utm_medium=organic&utm_source=githu
...
 * Rust runners, and therefore **Rust behaviors** (which are generally a subset of the @hash behaviors found within hIndex) are currently **disabled**. This will be taken on after we have enabled the Python runners. Similar to Python, a large amount of the implementation is finished and available to explore, but currently not-in-use.
 
-There are a number of other functionalities in the HASH platform that are possibly under-development and/or not stable within the current repository. Feel free to try things out, but don't be dissuaded if they don't work as of yet. We don't want to make any guarantees until we've had time to properly test features, and for now we're prioritising development to get those features out!
+There are a number of other functionalities in the HASH platform that are possibly under-development and/or not stable within the current repository. Feel free to try things out, but don't be dissuaded if they don't work yet. We don't want to make any guarantees until we've had time to properly test features, and for now we're prioritising development to get those features out!
```
Comment (Alfred-Mountfield) [inline packages/engine/README.md:58]: Also an unnecessary change in terms of grammar but it makes it easier to understand so we can keep it :)

File: packages/engine/README.md [hunk 3]
```diff
@@ -207,8 +206,7 @@ $ export RUST_LOG=debug
 
 
 ### Simulation Outputs
-> **WIP** - This section is a work-in-progress. More in-depth documentation is in the works for describing all output formats and options. As such some functionality may not be mentioned here, and some functionality alluded to here might not be complete at present. 
-Currently the engine has two main form of outputs, one coming from the [json_state package](./src/simulation/package/output/packages/json_state) and the other from the [analysis package](./src/simulation/package/output/packages/analysis).
+> **WIP** - This section is a work-in-progress. More in-depth documentation is in the works for describing all output formats and options. As such some functionality may not be mentioned here, and some functionality alluded to here might not be complete at present. Currently, the engine has two main form of outputs, one coming from the [json_state package](./src/simulation/package/output/packages/json_state) and the other from the [analysis package](./src/simulation/package/output/packages/analysis).
```
Comment (Alfred-Mountfield) [inline packages/engine/README.md:209]: > Currently, the engine has two main form of outputs, one coming from the [json_state package](./src/simulation/package/output/packages/json_state) and the other from the [analysis package](./src/simulation/package/output/packages/analysis).

Isn't part of the WIP disclaimer section

File: packages/engine/README.md [hunk 4]
```diff
@@ -28,8 +28,9 @@
   * [CLI Arguments and Options](#cli-arguments-and-options)
   * [Run a simulation](#run-a-simulation)
   * [Simulation Outputs](#simulation-outputs)
-    + [JSON-State](#json-state)
-    + [Analysis](#analysis)
+    + [JSON-State](#json-state-json_statejson)
```
Comment (nonparibus) [inline packages/engine/README.md:31]: What's going on here and on line 32?

File: packages/engine/README.md [hunk 5]
```diff
@@ -219,13 +217,83 @@ There are overrides ([CLI Arguments and Options](#cli-arguments-and-options)) fo
...
 
-During the run, the output may be buffered into the `./parts` folder in multiple files. These files are not necessarily valid JSON as the resultant state blob that appears within `json_state.json` is split up (hence `part`) for buffering purposes. 
+During the run, the output may be buffered into the `./parts` folder in multiple files. These files are not necessarily valid JSON as the resultant state blob that appears within `json_state.json` is split up (hence `part`) for buffering purposes.
 
 #### Analysis [`analysis_outputs.json`]
 > **WIP** - This feature is currently unstable
 
 [hCore] currently provides functionality where simulations can apply custom analysis on user-defined metrics. The functionality has been ported across to this codebase in the [analysis package](./src/simulation/package/output/packages/analysis), however development is planned to stabilise it. As such, this functionality is neither tested, nor considered supported.
 
+### Behavior keys
```
Comment (Alfred-Mountfield) [inline packages/engine/README.md:227]: I think this doesn't quite make sense as its own section. We should probably add a `### Simulation Inputs` section similar to `Simulation Outputs` and start describing the simulation project folder structure.

To avoid blowing up the scope of the PR though, we can avoid fleshing out all of it and just stick another WIP disclaimer

File: packages/engine/README.md (hunk 1 above)
Comment (nonparibus) [inline packages/engine/README.md:55]: Sentence still seems broken/reads strange to me. Appreciate that was the case prior to this PR as well.

**Review by nonparibus — APPROVED** [review]

**Comment by nonparibus** [comment]
Agree w/ Alfie's comments and added a couple more. Nothing breaking though.

File: packages/engine/README.md (hunk 4 above)
Comment (TimDiekmann) [reply packages/engine/README.md:31]: The headings are called <code>JSON-State [`json_state.json`]</code>, thus the extra "-json_statejson` is required.

File: packages/engine/README.md (hunk 1 above)
Comment (TimDiekmann) [reply packages/engine/README.md]: I updated the sentence, it should be more readable now.

File: packages/engine/README.md (hunk 4 above)
Comment (Alfred-Mountfield) [reply packages/engine/README.md:33]: Yeah just funky internal link specifics with how it strips characters and such

Comment (Alfred-Mountfield) [reply packages/engine/README.md:33]: Although that said, Tim are you sure the underscore stays?

Comment (Alfred-Mountfield) [reply packages/engine/README.md:33]: Weird, discussion on Github about this implies it should be stripped. But checked the branch and it works:
https://github.com/hashintel/hash/tree/td/behavior-readme/packages/engine#analysis-analysis_outputsjson

Good enough for me

File: packages/engine/README.md [hunk 6]
```diff
@@ -206,9 +206,81 @@ $ export RUST_LOG=debug
 [docs]: https://hash.ai/docs/simulation?utm_medium=organic&utm_source=github_readme_engine
 
 
+### Simulation Inputs
+> **WIP** - This section is a work-in-progress. More in-depth documentation is in the works for describing all input formats and options, and a project structure.
```
Comment (Alfred-Mountfield) [inline packages/engine/README.md:210]:
```suggestion
> **WIP** - This section is a work-in-progress. More in-depth documentation is in the works for describing all input formats and options, and expected project structure. For now, it's recommended that you create your simulations within [hCore] and use the "Export Project" functionality.
```

File: packages/engine/README.md [hunk 7]
```diff
@@ -206,9 +206,81 @@ $ export RUST_LOG=debug
...
+
+```json
+{
+  "keys": {},
+  "built_in_key_use": null,
+  "dynamic_access": true
+}
+```
+
+`"keys"` is a dictionary of state fields, every key corresponds to a field. Each field requires at least two keys: `"type"` and `"nullable"`. Depending on `"type"` other keys may be required. We support the following `"type"` values:
```
Comment (Alfred-Mountfield) [inline packages/engine/README.md:223]:
```suggestion
`"keys"` is a JSON object where each [key]/[value] pair is [behavior key name]/[behavior key specification] . The behavior key specification is a JSON object with at least two required members, `"type"` and `"nullable"`. Depending on `"type"` other members may be required. We support the following `"type"` values:
```

File: packages/engine/README.md [hunk 8]
```diff
@@ -206,9 +206,81 @@ $ export RUST_LOG=debug
 [docs]: https://hash.ai/docs/simulation?utm_medium=organic&utm_source=github_readme_engine
 
 
+### Simulation Inputs
+> **WIP** - This section is a work-in-progress. More in-depth documentation is in the works for describing all input formats and options, and a project structure.
+
+#### Behavior keys
+Behavior keys define the **data type** of the fields that a behavior accesses on an agent's state. See the [docs](https://hash.ai/docs/simulation/creating-simulations/behaviors/behavior-keys?utm_medium=organic&utm_source=github_readme_engine) for an explanation for behavior keys in general. If you don't have a project set up in [hCore], it's also possible to create the behaviors keys by hand. Generally, every state variable requires a key. The top level dictionary consist of up to three fields: `"keys"`, `"built_in_key_use"`, and `"dynamic_access"`, while the latter two are neither required, nor used currently:
```
Comment (Alfred-Mountfield) [inline packages/engine/README.md:213]:
```suggestion
Behavior keys define the fields, and their respective **data type**, that a behavior accesses on an agent's state. See the [docs](https://hash.ai/docs/simulation/creating-simulations/behaviors/behavior-keys?utm_medium=organic&utm_source=github_readme_engine) for an explanation of behavior keys in general. 

If you haven't created and exported a project from [hCore], it's also possible to manually create the file that specifies the behaviors keys. Generally, every user-defined variable on state (i.e. a behavior key) requires it to be specified within the accompanying `.json` file. The top level JSON object has up to three members, `"keys"`, `"built_in_key_use"`, and `"dynamic_access"`, while the latter two are neither required, nor used currently:
```

File: packages/engine/README.md [hunk 9]
```diff
@@ -206,9 +206,81 @@ $ export RUST_LOG=debug
...
+{
+  "keys": {},
+  "built_in_key_use": null,
+  "dynamic_access": true
+}
+```
+
+`"keys"` is a dictionary of state fields, every key corresponds to a field. Each field requires at least two keys: `"type"` and `"nullable"`. Depending on `"type"` other keys may be required. We support the following `"type"` values:
+
+- **`"any"`**: Can be any datatype (should be avoided if possible to improve performance)
```
Comment (Alfred-Mountfield) [inline packages/engine/README.md:225]:
```suggestion
- **`"any"`**: Can be any datatype (when performance becomes a concern, a specific data-type should be preferred)
```

File: packages/engine/README.md [hunk 10]
```diff
@@ -206,9 +206,81 @@ $ export RUST_LOG=debug
...
+- **`"number"`**: A 64 bit floating point number
+- **`"string"`**: The encoding depends on the language used for the behavior
+- **`"boolean"`**: Either `true` or `false`
+- **`"struct"`**: A nested object, additionally requires a `"fields"`-dictinary with the same schema as the top-level key `"keys"`. Example:
```
Comment (Alfred-Mountfield) [inline packages/engine/README.md:229]:
```suggestion
- **`"struct"`**: A nested object, which then additionally requires adding a new member called `"fields"` with the same schema as the top-level `"keys"`. Example:
```

File: packages/engine/README.md [hunk 11]
```diff
@@ -206,9 +206,81 @@ $ export RUST_LOG=debug
...
+            "field2": {
+              "type": "number",
+              "nullable": true
+            }
+          }
+        }
+      }
+    }
+    ```
+- **`"list"`**: an array with an abitirary number of subelements of the same type, which has to be specified with the `"child"` key:
```
Comment (Alfred-Mountfield) [inline packages/engine/README.md:250]:
```suggestion
- **`"list"`**: an array with an arbitrary number of sub-elements of the same type, which then have to be specified with the addition of another member called `"child"`:
```

File: packages/engine/README.md [hunk 12]
```diff
@@ -206,9 +206,81 @@ $ export RUST_LOG=debug
...
+          "type": "list",
+          "child": {
+            "type": "string",
+            "nullable": true
+          }
+        }
+      }
+    }
+    ```
+- **`"fixed-size-list"`**: an array with exactly `"length"` number of subelements of the same type, which has to be specified with the `"child"` key:
```
Comment (Alfred-Mountfield) [inline packages/engine/README.md:265]:
```suggestion
- **`"fixed-size-list"`**: an array with exactly `"length"` number of sub-elements of the same type, which then have to be specified with the addition of another member called `"child"`:
```

File: packages/engine/README.md [hunk 13]
```diff
@@ -206,9 +206,81 @@ $ export RUST_LOG=debug
...
+Behavior keys define the **data type** of the fields that a behavior accesses on an agent's state. See the [docs](https://hash.ai/docs/simulation/creating-simulations/behaviors/behavior-keys?utm_medium=organic&utm_source=github_readme_engine) for an explanation for behavior keys in general. If you don't have a project set up in [hCore], it's also possible to create the behaviors keys by hand. Generally, every state variable requires a key. The top level dictionary consist of up to three fields: `"keys"`, `"built_in_key_use"`, and `"dynamic_access"`, while the latter two are neither required, nor used currently:
+
+```json
+{
+  "keys": {},
+  "built_in_key_use": null,
+  "dynamic_access": true
+}
+```
+
```
Comment (Alfred-Mountfield) [inline packages/engine/README.md:222]:
```suggestion

> Note: JSON Objects have fields/members specified by [key]/[value] pairs. To avoid confusion between behavior _keys_ and the _keys_ in JSON Objects, we will be referring to the pairs as members or fields.

```

File: packages/engine/README.md [hunk 14]
```diff
@@ -206,9 +206,85 @@ $ export RUST_LOG=debug
...
+> Note: JSON Objects have fields/members specified by [key]/[value] pairs. To avoid confusion between behavior _keys_ and the _keys_ in JSON Objects, we will be referring to the pairs as members or fields.
+
+`"keys"` is a JSON object where each [key]/[value] pair is [behavior key name]/[behavior key specification] . The behavior key specification is a JSON object with at least two required members, `"type"` and `"nullable"`. Depending on `"type"` other members may be required. We support the following `"type"` values:
```
Comment (TimDiekmann) [inline packages/engine/README.md:227]:
```suggestion
`"keys"` is a JSON object where each [key]/[value] pair is [behavior key name]/[behavior key specification]. The behavior key specification is a JSON object with at least two required members, `"type"` and `"nullable"`. Depending on `"type"` other members may be required. We support the following `"type"` values:
```

**Review by Alfred-Mountfield — APPROVED** [review]
